[Clang] Remove unnecessary Decl transform & profiles for SizeOfPackExpr

zyn0217 opened the pull request:
We used to always transform the pattern declaration for SizeOfPackExpr to ensure the constraint expression's profile produced the desired result. However, this approach failed to handle pack expansions when the pack referred to function parameters. In such cases, the function parameters were formerly expanded to 1 to avoid building Subst* nodes (see e6974daa7). That workaround caused us to transform a pack without a proper ArgumentPackSubstitutionIndex, leading to crashes when transforming the pattern.

It turns out that profiling the pattern for partially substituted SizeOfPackExprs is unnecessary because their transformed forms are also profiled within the partial arguments.

(No release note because this is actually a trunk regression)

Fixes https://github.com/llvm/llvm-project/issues/124161

zyn0217 committed:
[Clang] Remove unnecessary Decl transform & profiles for SizeOfPackExpr

We used to always transform the pattern declaration for SizeOfPackExpr
to ensure the constraint expression's profile produced the desired
result. However, this approach failed to handle pack expansions when
the pack referred to function parameters. In such cases, the function
parameters were previously expanded to 1 to avoid building Subst* nodes
(see e6974daa7). That workaround caused us to transform a pack without
a proper ArgumentPackSubstitutionIndex, leading to crashes when
transforming the pattern.

It turns out that profiling the pattern for partially substituted
SizeOfPackExprs is unnecessary because their transformed forms are
already profiled within the partial arguments. Moreover, we always
end up with a partially transformed SizeOfPackExpr for constraint
comparison.

cor3ntin commented on clang/test/SemaTemplate/concepts-out-of-line-def.cpp:
```diff
@@ -720,6 +720,34 @@ template <typename... Ts> struct d {
 template struct c<int>;
 template struct d<int, int>;
 
+namespace Regression_123441 {
```
```suggestion
namespace GH123441 {
```

cor3ntin commented on clang/test/SemaTemplate/concepts-out-of-line-def.cpp:
```diff
@@ -720,6 +720,34 @@ template <typename... Ts> struct d {
+
 } // namespace GH115098
 
 namespace GH114685 {
```
Did you meant to be nested in another namespace ?

cor3ntin reviewed (approved):
LGTM except for nits

zyn0217 committed:
Address review feedback

mizvekov reviewed (approved):
LGTM, Thanks!

shafik reviewed (commented):
In my comment here, it really looks like a clang-17 regression: https://github.com/llvm/llvm-project/issues/124161#event-16079402729

From what I can tell you also address the simplified version, so I believe a release note is probably warrented.

Are there really multiple bugs here?

AaronBallman commented:
> In my comment here, it really looks like a clang-17 regression: [#124161 (comment)](https://github.com/llvm/llvm-project/issues/124161#event-16079402729)
> 
> From what I can tell you also address the simplified version, so I believe a release note is probably warrented.
> 
> Are there really multiple bugs here?

Yeah, this does appear to be a regression and so it warrants a release note. Reminder: please be sure to check back on issues to make sure we're addressing all of the comments (sometimes new information comes up in comments that's not reflected in the issue summary).

zyn0217 commented:
@shafik @AaronBallman Here is the complete backstory:

We were not performing well in out-of-line concept comparison, especially for expressions involving packs. I tried to address these issues around commit 04d20b170, but it quickly became evident that the fix was insufficient for https://github.com/llvm/llvm-project/issues/101735. To resolve that, we harnessed a feature that TreeTransform would expand function parameters if the parameters were instantiated as a pack, rather than keeping them in non-expanded forms.

That solution (developed during the 20 cycle and backported to 19) worked but introduced a pitfall: if we needed to transform a pack declaration within an expression, we would also need a valid ArgumentPackSubstitutionIndex to expand the pack.

Shortly after, I addressed #93099 (also during the 20 development cycle), where a SizeOfPackExpr was improperly profiled for constraint comparison due to its declaration being non-transformed. To that end, I added a declaration transformation step when normalizing constraints. Combined with the previously mentioned pitfall, this again crashed #124161 since no ArgumentPackSubstitutionIndex was set.

Technically, this should have worked at some point, as @shafik observed, which I believe was prior to the fix for #93099. And as far as I can tell, the latest crash was introduced by the fix for #93099, so I didn’t add a release note as usual.

AaronBallman commented:
Thank you for the explanation, that makes more sense to me.
